--- a/arch/arm/mach-ixp4xx/ixdp425-setup.c      2018-03-17 16:10:47.995685651 +1300
+++ b/arch/arm/mach-ixp4xx/ixdp425-setup.c      2018-03-14 19:52:43.192940814 +1300
@@ -30,6 +30,7 @@
 #include <asm/irq.h>
 #include <asm/mach/arch.h>
 #include <asm/mach/flash.h>
+#include <linux/usb/ehci_pdriver.h>

 #define IXDP425_SDA_PIN                7
 #define IXDP425_SCL_PIN                6
@@ -218,6 +219,65 @@
        &ixdp425_eth[1],
 };

+
+static struct resource kixrp435_usb0_resources[] = {
+       {
+               .start  = 0xCD000000,
+               .end            = 0xCD000300,
+               .flags  = IORESOURCE_MEM,
+       },
+       {
+               .start  = 32,
+               .flags  = IORESOURCE_IRQ,
+       },
+};
+
+static struct resource kixrp435_usb1_resources[] = {
+       {
+               .start  = 0xCE000000,
+               .end            = 0xCE000300,
+               .flags  = IORESOURCE_MEM,
+       },
+       {
+               .start  = 33,
+               .flags  = IORESOURCE_IRQ,
+       },
+};
+
+static u64 ehci_dma_mask = ~(u32)0;
+
+static struct usb_ehci_pdata kixrp435_usb_pdata = {
+       .big_endian_desc = 1,
+       .big_endian_mmio = 1,
+       .has_tt = 1,
+       .caps_offset = 0x100,
+};
+
+static struct platform_device kixrp435_usb0_device =  {
+       .name           = "ehci-platform",
+       .id             = 0,
+       .resource       = kixrp435_usb0_resources,
+       .num_resources  = ARRAY_SIZE(kixrp435_usb0_resources),
+       .dev = {
+               .dma_mask               = &ehci_dma_mask,
+               .coherent_dma_mask      = 0xffffffff,
+               .platform_data = &kixrp435_usb_pdata,
+       },
+};
+
+static struct platform_device kixrp435_usb1_device = {
+       .name           = "ehci-platform",
+       .id             = 1,
+       .resource       = kixrp435_usb1_resources,
+       .num_resources  = ARRAY_SIZE(kixrp435_usb1_resources),
+       .dev = {
+               .dma_mask               = &ehci_dma_mask,
+               .coherent_dma_mask      = 0xffffffff,
+               .platform_data = &kixrp435_usb_pdata,
+       },
+};
+
+
 static void __init ixdp425_init(void)
 {
        ixp4xx_sys_init();
@@ -246,6 +306,9 @@
        if (cpu_is_ixp43x()) {
                ixdp425_uart.num_resources = 1;
                ixdp425_uart_data[1].flags = 0;
+
+               platform_device_register(&kixrp435_usb0_device);
+               platform_device_register(&kixrp435_usb1_device);
        }

        platform_add_devices(ixdp425_devices, ARRAY_SIZE(ixdp425_devices));
@@ -310,4 +373,4 @@
        .dma_zone_size  = SZ_64M,
 #endif
 MACHINE_END
-#endif
+#endif
